Panic on weird n" in a Rust file

### Summary

I was writing the following line in Rust:

```
writeln!(writer, "{}\n", positions.len())?;
```

After entering `\n`, the editor crashed!

The panic is from the following line:

https://github.com/helix-editor/helix/blob/d8f9b901ddb1d91bf3ade66cf091e93b5cab6e49/helix-core/src/transaction.rs#L413

### Reproduction Steps

I found a minimal (and very weird) example to reproduce this bug:

1. Run the commands
    1. `cargo new project`
    2. `cd project`
    3. `echo 'n"' > src/main.rs`
    4. `hx src/main.rs`
2. Wait for the language server to start.
3. Press `c` over the character `n`  to change it.
4. Helix crashes with the following error:
    ```
    thread 'main' panicked at 'Positions [(1, After), (1, After)] are out of range for changeset len 2!', /var/home/mo/code/compile/helix/helix-core/src/transaction.rs:413:9
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
    ```

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-06-27T16:20:03.976 helix_view::document [WARN] LSP formatting failed: protocol error: ServerError(-32900): rustfmt exited with:
                           Status: exit status: 101
                           stdout: 
                           stderr: error: prefix `n` is unknown
 --> <stdin>:1:1
  |
1 | n"
  | ^ unknown prefix
  |
  = note: prefixed identifiers and literals are reserved since Rust 2021
help: consider inserting whitespace here
  |
1 | n "
  |  +

error[E0765]: unterminated double quote string
 --> <stdin>:1:2
  |
1 | n"
  |  ^^
```

</details>


### Platform

Linux

### Terminal Emulator

Alacritty 0.12.1

### Helix Version

helix 23.05 (d5111222)

--- a/helix-core/src/transaction.rs
+++ b/helix-core/src/transaction.rs
@@ -326,7 +326,7 @@ impl ChangeSet {
         self.changes.is_empty() || self.changes == [Operation::Retain(self.len)]
     }
 
-    /// Map a *sorted* list of positions through the changes.
+    /// Map a (mostly) *sorted* list of positions through the changes.
     ///
     /// This is equivalent to updating each position with `map_pos`:
     ///
@@ -335,27 +335,63 @@ impl ChangeSet {
     ///     *pos = changes.map_pos(*pos, assoc);
     /// }
     /// ```
-    /// However this function is significantly faster running in `O(N+M)` instead of `O(NM)`
+    /// However this function is significantly faster for sorted lists running
+    /// in `O(N+M)` instead of `O(NM)`. This function also handles unsorted/
+    /// partially sorted lists. However, in that case worst case complexity is
+    /// again `O(MN)`.  For lists that are often/mostly sorted (like the end of diagnostic ranges)
+    /// performanc is usally close to `O(N + M)`
     pub fn update_positions<'a>(&self, positions: impl Iterator<Item = (&'a mut usize, Assoc)>) {
         use Operation::*;
 
         let mut positions = positions.peekable();
-        macro_rules! map {
-            ($map: expr) => {
-                loop {
-                    let Some((pos, assoc)) = positions.peek_mut() else { return; };
-                    let Some(new_pos) = $map(**pos, *assoc) else { break; };
-                    **pos = new_pos;
-                    positions.next();
-                }
-            };
-        }
 
         let mut old_pos = 0;
         let mut new_pos = 0;
-        let mut iter = self.changes.iter().peekable();
+        let mut iter = self.changes.iter().enumerate().peekable();
 
-        while let Some(change) = iter.next() {
+        'outer: loop {
+            macro_rules! map {
+                ($map: expr, $i: expr) => {
+                    loop {
+                        let Some((pos, assoc)) = positions.peek_mut() else { return; };
+                        if **pos < old_pos {
+                            // Positions are not sorted, revert to the last Operation that
+                            // contains this position and continue iterating from there.
+                            // We can unwrap here since `pos` can not be negative
+                            // (unsigned integer) and iterating backwards to the start
+                            // should always move us back to the start
+                            for (i, change) in self.changes[..$i].iter().enumerate().rev() {
+                                match change {
+                                    Retain(i) => {
+                                        old_pos -= i;
+                                        new_pos -= i;
+                                    }
+                                    Delete(i) => {
+                                        old_pos -= i;
+                                    }
+                                    Insert(ins) => {
+                                        new_pos -= ins.chars().count();
+                                    }
+                                }
+                                if old_pos <= **pos {
+                                    iter = self.changes[i..].iter().enumerate().peekable();
+                                }
+                            }
+                            debug_assert!(old_pos <= **pos, "Reverse Iter across changeset works");
+                            continue 'outer;
+                        }
+                        let Some(new_pos) = $map(**pos, *assoc) else { break; };
+                        **pos = new_pos;
+                        positions.next();
+                    }
+                };
+            }
+
+            let Some((i, change)) = iter.next() else {
+                map!(|pos, _| (old_pos == pos).then_some(new_pos), self.changes.len());
+                break;
+            };
+
             let len = match change {
                 Delete(i) | Retain(i) => *i,
                 Insert(_) => 0,
@@ -364,42 +400,51 @@ impl ChangeSet {
 
             match change {
                 Retain(_) => {
-                    map!(|pos, _| (old_end > pos).then_some(new_pos + (pos - old_pos)));
+                    map!(
+                        |pos, _| (old_end > pos).then_some(new_pos + (pos - old_pos)),
+                        i
+                    );
                     new_pos += len;
                 }
                 Delete(_) => {
                     // in range
-                    map!(|pos, _| (old_end > pos).then_some(new_pos));
+                    map!(|pos, _| (old_end > pos).then_some(new_pos), i);
                 }
                 Insert(s) => {
                     let ins = s.chars().count();
 
                     // a subsequent delete means a replace, consume it
-                    if let Some(Delete(len)) = iter.peek() {
+                    if let Some((_, Delete(len))) = iter.peek() {
                         iter.next();
 
                         old_end = old_pos + len;
                         // in range of replaced text
-                        map!(|pos, assoc| (old_end > pos).then(|| {
-                            // at point or tracking before
-                            if pos == old_pos || assoc == Assoc::Before {
-                                new_pos
-                            } else {
-                                // place to end of insert
-                                new_pos + ins
-                            }
-                        }));
+                        map!(
+                            |pos, assoc| (old_end > pos).then(|| {
+                                // at point or tracking before
+                                if pos == old_pos || assoc == Assoc::Before {
+                                    new_pos
+                                } else {
+                                    // place to end of insert
+                                    new_pos + ins
+                                }
+                            }),
+                            i
+                        );
                     } else {
                         // at insert point
-                        map!(|pos, assoc| (old_pos == pos).then(|| {
-                            // return position before inserted text
-                            if assoc == Assoc::Before {
-                                new_pos
-                            } else {
-                                // after text
-                                new_pos + ins
-                            }
-                        }));
+                        map!(
+                            |pos, assoc| (old_pos == pos).then(|| {
+                                // return position before inserted text
+                                if assoc == Assoc::Before {
+                                    new_pos
+                                } else {
+                                    // after text
+                                    new_pos + ins
+                                }
+                            }),
+                            i
+                        );
                     }
 
                     new_pos += ins;
@@ -407,7 +452,6 @@ impl ChangeSet {
             }
             old_pos = old_end;
         }
-        map!(|pos, _| (old_pos == pos).then_some(new_pos));
         let out_of_bounds: Vec<_> = positions.collect();
 
         panic!("Positions {out_of_bounds:?} are out of range for changeset len {old_pos}!",)
@@ -822,6 +866,20 @@ mod test {
         };
         assert_eq!(cs.map_pos(2, Assoc::Before), 2);
         assert_eq!(cs.map_pos(2, Assoc::After), 2);
+        // unsorted selection
+        let cs = ChangeSet {
+            changes: vec![
+                Insert("ab".into()),
+                Delete(2),
+                Insert("cd".into()),
+                Delete(2),
+            ],
+            len: 4,
+            len_after: 4,
+        };
+        let mut positions = [4, 2];
+        cs.update_positions(positions.iter_mut().map(|pos| (pos, Assoc::After)));
+        assert_eq!(positions, [4, 2]);
     }
 
     #[test]

--- a/helix-core/src/transaction.rs
+++ b/helix-core/src/transaction.rs
@@ -339,7 +339,7 @@ impl ChangeSet {
     /// in `O(N+M)` instead of `O(NM)`. This function also handles unsorted/
     /// partially sorted lists. However, in that case worst case complexity is
     /// again `O(MN)`.  For lists that are often/mostly sorted (like the end of diagnostic ranges)
-    /// performanc is usally close to `O(N + M)`
+    /// performance is usally close to `O(N + M)`
     pub fn update_positions<'a>(&self, positions: impl Iterator<Item = (&'a mut usize, Assoc)>) {
         use Operation::*;
 

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -1197,7 +1197,7 @@ impl Document {
                 if let Some(data) = Rc::get_mut(annotations) {
                     changes.update_positions(
                         data.iter_mut()
-                            .map(|diagnostic| (&mut diagnostic.char_idx, Assoc::After)),
+                            .map(|annotation| (&mut annotation.char_idx, Assoc::After)),
                     );
                 }
             };

